CMakeLists.txt

```diff
@@ -20,7 +20,7 @@ set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/modules/")
 # Only GCC-like compilers support -Wextra, and other compilers give tons of
 # output for -Wall, so only -Wall and -Wextra on GCC.
 if(CMAKE_COMPILER_IS_GNUCC OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0 -Wall -Wextra -ftemplate-depth=1000 -pthread")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0 -Wall -Wextra -ftemplate-depth=1000 -pthread -fopenmp")
```

Comment:
Let's not include openmp by default, instead we should introduce another option, like:

https://github.com/mlpack/mlpack/blob/1b684afef5548c6795dc9ca3cb5f04eade23a14e/CMakeLists.txt#L86

Reply:
Sure, I will take a look and include a check like this. 
And can you help identify why the tests in the build pipeline are not working, I have not been able to make it work.

Reply:
It looks like all the tests are passing---maybe can you link to what's failing?  I'd be happy to take a look. :+1:

Reply:
Hi @rcurtin, If you look at the azure pipelines at the step where the tests are run you will see that only 1 test is running, and not all of them as on my local machine so I doubt there is something wrong with either the CMake file for catch2 or maybe the configuration that runs them on azure.

Below I am including the output on my local machine for a better idea. 

```
[WARN ] Unknown augmentation : 'horizontal-flip' not found!
[FATAL] Unknown Dataset. no-dataset For other datasets try loading data using generic dataloader functions such as LoadCSV. Refer to the documentation for more info.


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
models_test is a Catch v2.13.3 host application.
Run with -? for options

-------------------------------------------------------------------------------
CSVDataLoaderTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/dataloader_tests.cpp:24
...............................................................................

/home/aakash/models/tests/dataloader_tests.cpp:32: FAILED:
  REQUIRE( Utils::PathExists("./../data/iris.csv") == true )
with expansion:
  false == true

tar (child): /home/aakash/models/build/./../data/mnist.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
[FATAL] Corrupted Data for mnist downloaded.

-------------------------------------------------------------------------------
MNISTDataLoaderTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/dataloader_tests.cpp:60
...............................................................................

/home/aakash/models/tests/dataloader_tests.cpp:60: FAILED:
due to unexpected exception with message:
  fatal error; see Log::Fatal output

tar (child): /home/aakash/models/build/./../data/PASCAL-VOC-Test.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
[WARN ] The ./../data/PASCAL-VOC-Test/Annotations/ doesn't exist.

error: Col::subvec(): indices out of bounds or incorrectly used
-------------------------------------------------------------------------------
ObjectDetectionDataLoaderFieldTypeTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/dataloader_tests.cpp:92
...............................................................................

/home/aakash/models/tests/dataloader_tests.cpp:92: FAILED:
due to unexpected exception with message:
  Col::subvec(): indices out of bounds or incorrectly used

tar (child): /home/aakash/models/build/./../data/PASCAL-VOC-Test.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
[WARN ] The ./../data/PASCAL-VOC-Test/Uniform_Annotation/ doesn't exist.

error: Mat::rows(): indices out of bounds or incorrectly used
-------------------------------------------------------------------------------
ObjectDetectionDataLoaderMatTypeTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/dataloader_tests.cpp:180
...............................................................................

/home/aakash/models/tests/dataloader_tests.cpp:180: FAILED:
due to unexpected exception with message:
  Mat::rows(): indices out of bounds or incorrectly used

tar (child): /home/aakash/models/build/./../data/cifar-test.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
tar (child): /home/aakash/models/build/./../data/cifar-test.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
[WARN ] The ./../data/cifar-test/ doesn't exist.

error: Mat::rows(): indices out of bounds or incorrectly used
-------------------------------------------------------------------------------
LoadImageDatasetFromDirectoryTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/dataloader_tests.cpp:230
...............................................................................

/home/aakash/models/tests/dataloader_tests.cpp:230: FAILED:
due to unexpected exception with message:
  Mat::rows(): indices out of bounds or incorrectly used

-------------------------------------------------------------------------------
DownloadFileTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/utils_tests.cpp:19
...............................................................................

/home/aakash/models/tests/utils_tests.cpp:26: FAILED:
  REQUIRE( Utils::PathExists("./../data/iris.csv") == true )
with expansion:
  false == true

-------------------------------------------------------------------------------
CheckSumTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/utils_tests.cpp:36
...............................................................................

/home/aakash/models/tests/utils_tests.cpp:40: FAILED:
  REQUIRE( Utils::CompareCRC32("./../data/iris_test.csv", "3be1f79e") == true )
with expansion:
  false == true

[WARN ] Unable to create file for testing.
tar (child): /home/aakash/models/build/./../data/USCensus1990.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
-------------------------------------------------------------------------------
ExtractFilesTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/utils_tests.cpp:73
...............................................................................

/home/aakash/models/tests/utils_tests.cpp:81: FAILED:
  REQUIRE( Utils::PathExists("./../data/USCensus1990.csv") == true )
with expansion:
  false == true

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0Warning: Failed to create the file ./../data/test_image.jpg: No such file or 
Warning: directory
 13  6597   13   903    0     0    661      0  0:00:09  0:00:01  0:00:08   661
curl: (23) Failed writing body (0 != 903)
-------------------------------------------------------------------------------
CurlDownloadTest
-------------------------------------------------------------------------------
/home/aakash/models/tests/utils_tests.cpp:93
...............................................................................

/home/aakash/models/tests/utils_tests.cpp:104: FAILED:
  REQUIRE( Utils::PathExists("./../data/test_image.jpg") == true )
with expansion:
  false == true

===============================================================================
test cases: 15 |  6 passed | 9 failed
assertions: 58 | 49 passed | 9 failed
```

Reply:
Made it work for the azure pipeline too, you can take a look at the failing tests now in the builds too.

Reply:
All the failing tests are because of the inability to download datasets.

Reply:
I don't know the details of this repository.  Have you tried working with a debugger to figure out why the datasets are not being downloaded?  I think you may have to get your hands dirty to figure out what is going wrong here.  I took a brief look at the data loading code, and it looks like the URLs it should try to download from are fine.  You can see all the datasets on the mlpack website here: https://mlpack.org/datasets/.